addons/hr_timesheet/models/hr_timesheet.py

```diff
@@ -124,35 +124,78 @@ def _compute_department_id(self):
+        # 2/ Search all employees related to user_ids and employee_ids, in the selected companies
+        Employee = self.env['hr.employee']
+        employees = Employee.sudo().with_context(active_test=False).search([
+            '&', '|', ('user_id', 'in', user_ids), ('id', 'in', employee_ids), ('company_id', 'in', self.env.companies.ids)
+        ]) or Employee
+
+        #                 ┌─── active/in companies ────────> was found with... ─── employee_id ───> (A) There is nothing to do, we will use this employee_id
+        # 3/ Each employee                                                    └─── user_id ───────> (B)** We'll need to select the right employee for this user
+        #                 └─── archived/not in companies ──> (C) We raise an error as we can't create a timesheet for an archived employee
+        # ** We can rely on the user to get the employee_id if
+        #    he has an active employee in the company of the timesheet
+        #    or he has only one active employee for all selected companies
+        active_employee_ids_in_vals = []
+        employee_per_company_per_user = defaultdict(dict)
         for employee in employees:
-            employee_for_user_company[employee.user_id.id][employee.company_id.id] = employee.id
-
-        employee_ids = set()
+            if employee.active:
+                if employee.id in employee_ids:
+                    active_employee_ids_in_vals.append(employee.id)
+                else:
+                    employee_per_company_per_user[employee.user_id.id][employee.company_id.id] = employee.id
+
+        # 4/ Put valid employee_id in each vals
+        error_msg = _('Timesheets must be created with an active employee in the selected companies.')
         for vals in vals_list:
-            # compute employee only for timesheet lines, makes no sense for other lines
-            if not vals.get('employee_id') and vals.get('project_id'):
-                employee_for_company = employee_for_user_company.get(vals.get('user_id', default_user_id), False)
-                if not employee_for_company:
-                    continue
-                company_id = list(employee_for_company)[0] if len(employee_for_company) == 1 else vals.get('company_id', self.env.company.id)
-                vals['employee_id'] = employee_for_company.get(company_id, False)
-            elif vals.get('employee_id'):
-                employee_ids.add(vals['employee_id'])
-        if any(not emp.active for emp in self.env['hr.employee'].browse(list(employee_ids))):
-            raise UserError(_('Timesheets must be created with an active employee.'))
+            if not vals.get('project_id'):
+                continue
+            user_id = False
+            employee_in = vals.get('employee_id')
+            if employee_in:
+                if employee_in in active_employee_ids_in_vals:
+                    vals['user_id'] = Employee.browse(employee_in).user_id.id
```

Comment:
We didn't have to prefetch the user in sudo in the case the user is just a timesheet user? 🤔